service/matching/matcher_test.go

```diff
@@ -209,55 +209,67 @@ func (t *MatcherTestSuite) testRemoteSyncMatch(taskSource enumsspb.TaskSource) {
 	t.Equal(mustParent(t.queue.partition.(*tqid.NormalPartition), 20).RpcName(), req.GetTaskQueue().GetName())
 }
 
+// validateSyncMatchWhenNoBacklog is a helper which validates a sync match when there is no backlog
+func (t *MatcherTestSuite) validateSyncMatchWhenNoBacklog(ctx context.Context, historyTask *internalTask) {
+	go func() {
+		ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+		_, err := t.rootMatcher.Poll(ctx, &pollMetadata{})
+		t.Assert().NoError(err)
+		cancel()
+	}()
+	time.Sleep(2 * time.Millisecond) //nolint:forbidigo
+
+	go func() {
+		historyTask.responseC <- nil
+	}()
+	time.Sleep(2 * time.Millisecond) //nolint:forbidigo
+	happened, err := t.rootMatcher.Offer(ctx, historyTask)
+	t.True(happened)
+	t.Nil(err)
+}
+
 func (t *MatcherTestSuite) TestRejectSyncMatchWhenBacklog() {
 	historyTask := newInternalTaskForSyncMatch(randomTaskInfo().Data, nil)
-	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
 
 	// sync match happens when there is no backlog
-	t.client.EXPECT().PollWorkflowTaskQueue(gomock.Any(), gomock.Any(), gomock.Any()).Return(&matchingservice.PollWorkflowTaskQueueResponse{}, errMatchingHostThrottleTest)
-	go t.childMatcher.Poll(ctx, &pollMetadata{}) //nolint:errcheck
-	time.Sleep(time.Millisecond)
-	go func() { historyTask.responseC <- nil }()
-	happened, err := t.childMatcher.Offer(ctx, historyTask)
-	t.True(happened)
-	t.Nil(err)
+	t.validateSyncMatchWhenNoBacklog(ctx, historyTask)
 
+	// first task waits for a local poller
 	intruptC := make(chan struct{})
 	youngBacklogTask := newInternalTaskFromBacklog(randomTaskInfoWithAge(time.Second), nil)
-	t.client.EXPECT().AddWorkflowTask(gomock.Any(), gomock.Any(), gomock.Any()).Return(&matchingservice.AddWorkflowTaskResponse{}, errMatchingHostThrottleTest)
-	go t.childMatcher.MustOffer(ctx, youngBacklogTask, intruptC) //nolint:errcheck
+	go t.rootMatcher.MustOffer(ctx, youngBacklogTask, intruptC) //nolint:errcheck
 	time.Sleep(time.Millisecond)
 
 	// should allow sync match when there is only young tasks in the backlog
-	t.client.EXPECT().AddWorkflowTask(gomock.Any(), gomock.Any(), gomock.Any()).Return(&matchingservice.AddWorkflowTaskResponse{}, errMatchingHostThrottleTest)
-	happened, err = t.childMatcher.Offer(ctx, historyTask)
+	t.client.EXPECT().AddWorkflowTask(gomock.Any(), gomock.Any(), gomock.Any()).Do(
+		func(arg0 context.Context, arg1 *matchingservice.AddWorkflowTaskRequest, arg2 ...interface{}) {
+		},
+	).Return(&matchingservice.AddWorkflowTaskResponse{}, errMatchingHostThrottleTest)
+	happened, err := t.childMatcher.Offer(ctx, historyTask)
```

Comment:
Offering on the child matcher to just show that we don't receive a return value of `false` because of a non-negligeable backlog and we actually conduct forwarding of a task. If we changed this to the root matcher, we simply would get a `false` return value since we don't have a poller polling at this point.

Reply:
why not just have an check that Offer returns true? Also the oldBacklog thing is not tested properly if there is no poller (and no forwarding) because it would return false anyway, right?

Reply:
Offer will return `true` only if the task is matched with a local poller. In this case, we don't have a local poller *yet*.  If we start a local poller, the initial task that is blocked(which gets polled later on) will be matched. The current task is being offered *just* to show that a task, if young, will get sync matched. I had thought of removing this excerpt but I think we should include this to showcase that only old tasks when offered will be rejected.

Now, reading your latter part of the comment made me realize - why don't we just have:

1. One `MustOffer` on the rootMatcher which shall make the backlog non-negligeable.
2. `Offer` on rootMatcher which shall reject sync match cause of point number 1.

I think this makes the test much much simpler - If this is what you had intended before, my apologies for not catching on to this until now. I thought your point was simply changing child matcher to root matcher for the test.

Reply:
I spent some more time thinking about if we can make this case simpler than the current implementation or not - while we certainly can, I am recommending we keep it to the current version because:

1. I think the current test case correctly shows what happens when we `Offer` tasks while having a young and an old backlog. This is important to show that we only reject sync match when we have older tasks in the backlog.

Let me know what you think.

Reply:
Makes sense, let's use the child matcher then, and I think you need to bring back your locking stuff. Please add comments in the code explaining why it's needed to use childMatcher and also what race condition the locking is protecting from.

Reply:
validateSyncMatchWhenNoBacklog can keep using the root matcher and become a separate test altogether.

Reply:
"bring back your locking stuff" - not sure I understand this part. The locking stuff is not required now that we are using the root matcher. It is only required when we use a child matcher, due to the race condition, and I am using that in the event where a forwarding of the task is attemped.

According to me, the latest commit which uses the root matcher (for the most part), is the right implementation of this test case. Did you also mean this?

Reply:
Oh, now the young task is `MustOffer`ed to the root partition by we are testing sync match on child. That does not test what this part of the test wanted to test, it wants to MustOffer the young task to the same partition as the sync match is tested so we make sure a non-zero but negligible backlog age does not block sync match.

Now that thinking more about it, it's just fine to test the old task case and ignore the young case. It is tested in other tests that a young task behaved correctly when it comes to backlog age and I don't think we need to test it again specifically for sync match behavior.